### ISSUE TITLE
StackTrace() method of LogEventInfo class of NLog giving null value when upgrade to NLog version 4.5.4

### ISSUE THREAD

**Issue by RudraRavi** [NLog/NLog#2721]
Hi! Thanks for reporting this feature/bug/question! 

Please keep / fill in the relevant info from this template so that we can help you as best as possible.

QUESTIONS are preferred on StackOverflow. You could expect a faster response there (but do include all the info below). Use the tag "nlog" https://stackoverflow.com/questions/ask

For .NET Core users, please check the platform support: https://github.com/NLog/NLog/wiki/platform-support

**Type** (choose one):

- Bug
- Feature request
- Question

**NLog version**: (e.g. 4.2.3)

**Platform**: .Net 3.5 / .Net 4.0 / .Net 4.5 / Mono 4 / Silverlight 4 / Silverlight 5 / Xamarin Android / Xamarin iOs / .NET Core 

**Current NLog config** (xml or C#, if relevant)

```xml
<nlog>
  <targets>
  </targets>
  <rules>
  </rules>
</nlog>

```

In case of a BUG:

- What is the current result?
- What is the expected result?
- Did you checked the [Internal log](https://github.com/NLog/NLog/wiki/Internal-Logging)?
- Please post full exception details (message, stacktrace, inner exceptions)
- Are there any workarounds? yes/no
- Is there a version in which it did work? 
- Can you help us by writing an unit test?

In case of a FEATURE REQUEST:
 
 - Why do we need it?
 - An example of the XML config, if relevant.

**Comment by RudraRavi** [comment]
What is the current result? : LogEventInfo.StackTrace() gives null value
What is the expected result? : this should return stack trace
Did you checked the Internal log?: yes
Please post full exception details (message, stacktrace, inner exceptions) : it is null reference exception as we are calculating GetFrame() afterward and due to null value of stacktrace it is giving object reference set to null.
Are there any workarounds? yes/no : not sure
Is there a version in which it did work? : Yes in 4.4.5
Can you help us by writing an unit test? : Yes

**Comment by 304NotModified** [comment]
@RudraRavi please share (stripped version of) your config. Thanks!

**Comment by RudraRavi** [comment]
We have custom Target class inherit from Target class of NLog .The only problem we are facing after using latest version of NLog that is null value of LogEventInfo.StackTrace(). This is not null for older version till NLog version 4.4.5. I am not using nlog config as handled in code for target.

**Comment by snakefoot** [comment]
@RudraRavi Could you share the code for your custom Target ? How do you signal that you want to enable capture of StackTrace ?

Normally one needs to perform a trick with revealing an object that inherits from `IUsesStackTrace`. Like this where they use public instance of CallSiteLayoutRenderer.

https://github.com/GibraltarSoftware/Gibraltar.Agent.NLog2/blob/master/src/Agent.NLog2/GibraltarTarget.cs

**Comment by snakefoot** [comment]
Funny looks like #1310 actually breaks the "trick" used by the GibraltarTarget, but that regression was introduced with NLog 4.3 (Older version used here).

#2048 tries to improve on this regression, but doesn't solve it for GibraltarTarget. Included with NLog 4.4.6.

@RudraRavi Does the StackTrace work when using NLog ver. 4.4.13 ?

**Comment by snakefoot** [comment]
If the following lines was added to Target.FindAllLayouts(), then it will resolve the regression from #1310 completely:

```c#
StackTraceUsage = _allLayouts.DefaultIfEmpty().Max(layout => layout == null ? StackTraceUsage.None : layout.StackTraceUsage);
// Scan for any random LayoutRenderer objects
if (StackTraceUsage != StackTraceUsage.Max)
    StackTraceUsage = ObjectGraphScanner.FindReachableObjects<IUsesStackTrace>(false, this).DefaultIfEmpty().Max(layout => layout == null ? StackTraceUsage.None : layout.StackTraceUsage);
```

**Comment by snakefoot** [comment]
Created #2727 to make it easier for custom targets to request capture of StackTrace. "Just" have to implement `NLog.Internal.IUsesStackTrace`

**Comment by 304NotModified** [comment]
> implement NLog.Internal.IUsesStackTrace

I think that's the best solution. The support for custom targets with stacktrace is then really supported.

**Comment by 304NotModified** [comment]
IMO this is fixed in 4.5.5, so close.
